lib/Basics/Thread.cpp

```diff
@@ -197,8 +197,29 @@ void Thread::shutdown() {
       // we must ignore any errors here, but TRI_DetachThread will log them
       TRI_DetachThread(&_thread);
     } else {
-      // we must ignore any errors here, but TRI_JoinThread will log them
-      TRI_JoinThread(&_thread);
+      auto timeout = 5 * 60 * 1000; // wait max 5min for the thread to terminate
```

Comment:
That is undesired behavior for every single thread in ArangoDB inherited from `arangodb::Thread` to throw an error after hard-coded period of time. That may break current logic. Please make timeout configurable with `INFINITE` value by default.

Reply:
Before the refactoring in #8032 this used to be the default behaviour, i.e., it used to be the case that every thread inherited from `arangodb::Thread` would cause process abortion if it did not stop within a timely manner after receiving the stop signal.
This PR tries to re-establish this behaviour, but of course it can be extended to be configurable. Can you elaborate why you think that this might break current logic?

Reply:
I bet that was done with the assumption that the only consumers of `arangodb::Thread` are descendants related to replication, cluster, etc. For that part of ArangoDB hard-coded timeout is probably fine. But since `arangodb::Thread` is generic, it may be used for different purposes, that's unexpected, odd and dangerous behavior to terminate *the whole database process* after `hard-coded` 5 min timeout.
Since there is no good default value for everyone, we have to:
- avoid unexpected termination of the process by using `INFINITE` default value 
- let each consumer decide how long one wants to wait

Reply:
Alright, I introduced a virtual method `getTerminationTimeout`. The default return value is 5min to be consistent with the old behaviour, but derived thread classes can override this method to customize the behaviour. A value of -1 (=`INFINITE`) indicates that we want to wait forever.

Reply:
@mpoeter thank you! May I ask you to introduce a member variable one can set in constructor instead of virtual method. Since `Thread::shutdown` is meant to be called from destructor of the derived type, having using virtual getter is error-prone.
@jsteemann can we make a final decision on default value. I would say it has to be `INFINITE` since it guarantees process not to be terminated unexpectedly. But in that case we have to adjust all necessary threads to set a suitable timeout explicitly

Reply:
@gnusi I am aware of the potential issues with virtual function calls in the destructor, but this also applies to `beginShutdown` (which is also called inside `shutdown`). And that is exactly the reason why I added the comment _MUST be called from the destructor of the MOST DERIVED class_ to the `shutdown` method and adapted all thread implementations to follow this rule. If this rule is broken we have much worse problems than `getTerminationThread` not returning the correct value! (This is the reason thy I did the refactorings in #8032 in the first place.)

IMHO a virtual function communicates more clearly that a thread deviates from the default behaviour, since the override has to be defined in the class declaration.
Regarding the default value: this is not my place to make such a decision - I just wanted to restore the previous behaviour. If you decide to change the default behaviour I am happy to make the according adaptions.

Reply:
Ok, my opinion here is: let's make the timeout parameter a parameter of the Thread constructor. This way it is explicit. And when a thread is created in a new piece of code, the developer has to explicitly set a timeout... If it's a virtual method of Thread with a default implementation, it is too easily forgotten to override the timeout value method `getTerminationTimeout`.

An alternative is to make `getTerminationTimeout` pure virtual, so it *has* to be overridden.
Then we only need a new class derived from Thread which implements `getTerminationTimeout` and returns 5 minutes, e.g. `ThreadWithDefaultTimeout`. And all existing Threads can then derive from this new class except the ones where we want different behavior.

I could live with both solutions.
WDYT?

Reply:
I can also live with both solutions. This is your code, so I am happy to oblige to a majority decision. :slightly_smiling_face:

Reply:
I would prefer to have a parameter of the `Thread` constructor, IMHO virtual getter is over-complication of simple things in that case. As I mentioned before, `INFINITE` is the only possible default value for that parameter, otherwise one may unexpectedly terminate the process.

Reply:
Alright, I will adapt the implementation accordingly.

Reply:
@mpoeter : thank you so much!

Reply:
@jsteemann no problem - my changes ignited this whole discussion in the first place. :slightly_smiling_face:
I am happy to change the default value to `INFINITE`, but I do need some input from your side about which threads might need a different timeout value. I know that @dothebart was struggling with some tests where a thread did not terminate due to some deadlock, so he in this scenario he would have needed the old behaviour.